src/Compilers/CSharp/Test/Emit2/Semantics/RefReadonlyParameterTests.cs

```diff
@@ -1251,4 +1251,691 @@ static class C
+    {
+        var source = """
+            class C
+            {
+                static void M(ref readonly int p) => System.Console.WriteLine(p);
+                static unsafe void Main()
+                {
+                    delegate*<ref readonly int, void> f = &M;
+                    int x = 111;
+                    f(ref x);
+                }
+            }
+            """;
+        var verifier = CompileAndVerify(source, expectedOutput: "111", options: TestOptions.UnsafeReleaseExe, verify: Verification.Fails);
+        verifier.VerifyDiagnostics();
+        verifier.VerifyMethodBody("C.Main", """
+            {
+              // Code size       19 (0x13)
+              .maxstack  2
+              .locals init (int V_0, //x
+                            delegate*<ref readonly int, void> V_1)
+              // sequence point: delegate*<ref readonly int, void> f = &M;
+              IL_0000:  ldftn      "void C.M(ref readonly int)"
+              // sequence point: int x = 111;
+              IL_0006:  ldc.i4.s   111
+              IL_0008:  stloc.0
+              // sequence point: f(ref x);
+              IL_0009:  stloc.1
+              IL_000a:  ldloca.s   V_0
+              IL_000c:  ldloc.1
+              IL_000d:  calli      "delegate*<ref readonly int, void>"
+              // sequence point: }
+              IL_0012:  ret
+            }
+            """);
+    }
+
+    [Fact]
+    public void Invocation_Delegate()
+    {
+        var source = """
+            delegate void D(ref readonly int p);
+            class C
+            {
+                static void M(ref readonly int p) => System.Console.WriteLine(p);
+                static void Main()
+                {
+                    D d = M;
+                    int x = 111;
+                    d(ref x);
+                }
+            }
+            """;
+        var verifier = CompileAndVerify(source, expectedOutput: "111");
+        verifier.VerifyDiagnostics();
+        verifier.VerifyMethodBody("C.Main", """
+            {
+              // Code size       38 (0x26)
+              .maxstack  2
+              .locals init (int V_0) //x
+              // sequence point: D d = M;
+              IL_0000:  ldsfld     "D C.<>O.<0>__M"
+              IL_0005:  dup
+              IL_0006:  brtrue.s   IL_001b
+              IL_0008:  pop
+              IL_0009:  ldnull
+              IL_000a:  ldftn      "void C.M(ref readonly int)"
+              IL_0010:  newobj     "D..ctor(object, System.IntPtr)"
+              IL_0015:  dup
+              IL_0016:  stsfld     "D C.<>O.<0>__M"
+              // sequence point: int x = 111;
+              IL_001b:  ldc.i4.s   111
+              IL_001d:  stloc.0
+              // sequence point: d(ref x);
+              IL_001e:  ldloca.s   V_0
+              IL_0020:  callvirt   "void D.Invoke(ref readonly int)"
+              // sequence point: }
+              IL_0025:  ret
+            }
+            """);
+    }
```

Comment:
>} [](http://example.com/codeflow?start=4&length=1)

We are verifying IL for `ref` used at the call site. We should also verify IL for cases when `in` is used and no modifier is used, including rvalue scenario.
It would be good to test what happens when `in` parameter is passed to `ref readonly` parameter, when  `ref readonly` parameter is passed to `in` parameter, and when  `ref readonly` parameter is passed to `ref readonly` parameter. With whatever modifiers we support at the call site.
Also it would be good to test a scenario when a readonly field is passed to `ref readonly` parameter outside of constructor.

Reply:
Thanks for the suggestions, will add. I have some rvalue tests in a follow-up PR which relaxes value checks for `ref readonly` parameters; I will add them here, but some won't yet produce the expected results (I will include prototype comments for those).